### ISSUE TITLE
Weird bug when using Cython

### ISSUE THREAD

**Issue by aytekinar** [tox-dev/tox#1489]
I have attached the project archive for reproducibility purposes. In short, I am running on OSX with the following
```bash
~> python3 --version
Python 3.7.5
```

Once you download the project archive (`tox-report.zip`), simply run the following (outputs trimmed):
```bash
~> unzip tox-report.zip
~> cd tox-report
~/tox-report> python3 -m venv venv
~/tox-report> source venv/bin/activate
~/tox-report> pip install --upgrade cython tox
~/tox-report> python setup.py build # to generate the .c file
~/tox-report> tox
```

The command fails with
```bash
ModuleNotFoundError: No module named 'mylib.mydot_cy'
```

However, the first two lines of `tox` output are as follows:
```bash
.package create: /Users/eaytard/tox-report/.tox/.package
.package installdeps: setuptools>=42.0.2, wheel>=0.33.6, cython>=0.29.14
```
which indicates that `tox` successfully parses `pyproject.toml`. Moreover, when I check the environment's site packages, I see the following:
```bash
~/tox-report> ls -al .tox/py37/lib/python3.7/site-packages/mylib
total 1824
drwxr-xr-x   9 eaytard  staff     288 Dec 24 00:16 .
drwxr-xr-x  17 eaytard  staff     544 Dec 24 00:16 ..
-rw-r--r--   1 eaytard  staff       0 Dec 24 00:16 __init__.py
drwxr-xr-x   5 eaytard  staff     160 Dec 24 00:16 __pycache__
-rw-r--r--   1 eaytard  staff      99 Dec 24 00:16 mydot.py
-rw-r--r--   1 eaytard  staff  750544 Dec 24 00:16 mydot_cy.c
-rwxr-xr-x   1 eaytard  staff  166780 Dec 24 00:16 mydot_cy.cpython-37m-darwin.so
-rw-r--r--   1 eaytard  staff      88 Dec 24 00:16 mydot_cy.pxd
-rw-r--r--   1 eaytard  staff     146 Dec 24 00:16 mydot_py.py
```

Clearly, `python setup.py sdist` packages the library properly with the built C-files from `python setup.py build`. I can also see the shared object in place.

I simply do not understand why my tests fail with the `ModuleNotFoundError` exception. Am I missing something very trivial, or do you think this might be a bug?

**NOTE.** If I instead run
```bash
~> unzip tox-report.zip
~> cd tox-report
~/tox-report> python3 -m venv venv
~/tox-report> source venv/bin/activate
~/tox-report> pip install --upgrade tox
~/tox-report> pip install -e .
~/tox-report> tox
```
the tests pass. I believe that this has something to do with proper merging of different environments.

[tox-report.zip](https://github.com/tox-dev/tox/files/3996546/tox-report.zip)

**Comment by gaborbernat** [comment]
If you're using your mylib from within setup.py I must tell you that per pep-517 build frontends such as tox are not required to put the current working directory onto the python path, and as such importing your own package will not work. If you import your own package ensure you put the setup.py working directory onto the path.

**Comment by gaborbernat** [comment]
Note in a pep 517 world setuptools is no longer called via setup.py invocations, so the setup.py folder is no longer automatically put onto the python path for the import system. In such cases you should start your setup script with altering the sys.path before doing the import of the mylib.

**Comment by aytekinar** [comment]
**TL;DR.** This issue _might_ still a false positive.

Hi @gaborbernat. First of all, I would like to thank you for your time and prompt response. I am not sure if you checked `tox-report.zip`, or you simply did not have any time to have a look, but I have read all your blog post series (thanks, by the way, for the nice explanations) as well as the corresponding PEP's to make sure I was not doing anything wrong/fishy.

I am a (_rather_) new developer in Python, and I was experimenting with how to best structure the source tree and initiate a nice CI/CD pipeline for some Python library. In my example, I was not importing `mylib` in `setup.py` in any way, either. The problem turned out to be the fact that I had the following source tree:

```bash
~/tox-report> tree -L 2 .
.
├── README.md
├── mylib
│   ├── __init__.py
│   ├── mydot.py
│   ├── mydot_cy.pxd
│   ├── mydot_cy.pyx
│   └── mydot_py.py
├── pyproject.toml
├── setup.py
├── test.py
├── tests
│   ├── __init__.py
│   ├── test_mydot.py
│   ├── test_mydot_cy.py
│   └── test_mydot_py.py
└── tox.ini

2 directories, 14 files
```
and the following lines in my `tox.ini` file:
```bash
~/tox-report> cat tox.ini 
# tox (https://tox.readthedocs.io/) is a tool for running tests
# in multiple virtualenvs. This configuration file will run the
# test suite on all supported python versions. To use it, "pip install tox"
# and then run "tox" from this directory.

[tox]
envlist = py37
isolated_build = True

[testenv]
deps =
    coverage
commands =
    coverage run --source mylib test.py
    coverage report
```
The `coverage` commands were run under `~/tox-report`, which resulted in the directory (package) `mylib`'s shadowing the isolated version under `.tox`. Even though `tox` successfully parsed `pyproject.toml` file, installed build dependencies and created a nice wheel distribution for `mylib`, the tests failed complaining about the C extension not being available under `mylib`. This was solved when I `pip install --no-deps -e .`, which puts the compiled C extension under `mylib` directory. However, this is against the idea behind having isolated builds, in my opinion.

Later, I changed the source tree to have `mylib` under another directory, `src`. Doing the necessary modifications to `setup.py` so that `setuptools` searches for the packages under `src`, my problem got solved. This time, because `mylib` directory does not exist (as a first level) under the source tree, the package inside the isolated build was used (i.e., the one under `.tox`).

Changing the `coverage` line to read `coverage run -m unittest` and getting rid of `test.py` does not solve the issue, either. Still, the import lines in the test files under `tests` read as `from mylib import X`, which results in the same problem. Do you think this should be the way to structure the source tree?

As a side question... You have stated
> Note in a pep 517 world setuptools is no longer called via setup.py invocations

I believe that this is true when installing or testing the libraries. To package the library, I still need to use `python setup.py {build,sdist,bdist_wheel}`, right? I mean, I am aware of `pip wheel --no-deps -w dist .`, which creates a wheel distribution, but because `pip` does not currently have `pip sdist` or `pip build` type of functionality, I need to use `sdist` directly from `setup.py`, right?

**Comment by gaborbernat** [comment]
> I believe that this is true when installing or testing the libraries. To package the library, I still need to use python setup.py {build,sdist,bdist_wheel}, right? I mean, I am aware of pip wheel --no-deps -w dist ., which creates a wheel distribution, but because pip does not currently have pip sdist or pip build type of functionality, I need to use sdist directly from setup.py, right?

You should be using ``pep517.build`` via https://github.com/pypa/pep517
